### ISSUE TITLE
Crash DDLog.m line 839 - [DDLog lt_log:]

### ISSUE THREAD

**Issue by andrey-filin-arlo** [CocoaLumberjack/CocoaLumberjack#1219]
### New Issue Checklist

- [x] I have read and understood the [CONTRIBUTING guide](https://github.com/CocoaLumberjack/CocoaLumberjack/blob/master/.github/CONTRIBUTING.md)
- [x] I have read the [Documentation](https://github.com/CocoaLumberjack/CocoaLumberjack#documentation) or [CocoaDocs](cocoadocs.org/docsets/CocoaLumberjack)
- [x] I have searched for a similar issue in the [project](https://github.com/CocoaLumberjack/CocoaLumberjack/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name                      | iOS
 Platform Version                   | 12.1.2
 CocoaLumberjack Version   | 3.6.0
 Integration Method               |  cocoapods
 Xcode Version                       | Xcode 12.2
 Repro rate                              | sometimes 67%

```
### Stack:
cocoa.lumberjack
0  libsystem_kernel.dylib         0x1a85539fc __ulock_wait + 8
1  libdispatch.dylib              0x1a8397238 _dispatch_ulock_wait + 56
2  libdispatch.dylib              0x1a8397c38 _dispatch_group_wait_slow + 40
3  CocoaLumberjack                0x104adae78 -[DDLog lt_log:] + 839 (DDLog.m:839)
4  CocoaLumberjack                0x104ad962c __40-[DDLog queueLogMessage:asynchronously:]_block_invoke + 358 (DDLog.m:358)
5  libdispatch.dylib              0x1a83f56c8 _dispatch_call_block_and_release + 24
6  libdispatch.dylib              0x1a83f6484 _dispatch_client_callout + 16
7  libdispatch.dylib              0x1a839dbd0 _dispatch_lane_serial_drain$VARIANT$mp + 592
8  libdispatch.dylib              0x1a839e718 _dispatch_lane_invoke$VARIANT$mp + 432
9  libdispatch.dylib              0x1a83a6eb8 _dispatch_workloop_worker_thread + 600
10 libsystem_pthread.dylib        0x1a85d90dc _pthread_wqthread + 312
11 libsystem_pthread.dylib        0x1a85dbcec start_wqthread + 4
```

**Comment by andrey-filin-arlo** [comment]
The issue is still relevant, we still have a lot of similar crashes

**Comment by ffried** [comment]
Do you have some more stack traces? Possibly also including other threads? It _looks_ like a deadlock, but for that we'd need to see where the call is coming from.

**Comment by silverhammermba** [comment]
I am seeing this crash as well. Here is one kind of stack trace we see

```
Exception Type:  EXC_CRASH (SIGKILL)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Triggered by Thread:  0


Thread 0 name:
Thread 0 Crashed:
0   libsystem_kernel.dylib        	0x00000001bc6a924c kevent_id + 8
1   libdispatch.dylib             	0x0000000182ddbfb4 _dispatch_kq_poll + 228 (event_kevent.c:750)
2   libdispatch.dylib             	0x0000000182ddc9f0 _dispatch_event_loop_wait_for_ownership$VARIANT$mp + 440 (event_kevent.c:2221)
3   libdispatch.dylib             	0x0000000182dcb820 __DISPATCH_WAIT_FOR_QUEUE__ + 288 (queue.c:1678)
4   libdispatch.dylib             	0x0000000182dcb464 _dispatch_sync_f_slow + 136 (queue.c:1762)
5   Logging                       	0x00000001020f3560 -[DDLog queueLogMessage:asynchronously:] + 95584 (DDLog.m:336)
6   Logging                       	0x00000001020f3b84 -[DDLog log:message:level:flag:context:file:function:line:tag:] + 97156 (DDLog.m:476)
7   Logging                       	0x00000001020f3784 -[DDLog log:level:flag:context:file:function:line:tag:format:] + 96132 (DDLog.m:0)
```

And another

```
Exception Type:  EXC_CRASH (SIGKILL)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Triggered by Thread:  0


Kernel Triage:
VM - pmap_enter failed with resource shortage


Thread 0 name:
Thread 0 Crashed:
0   libsystem_kernel.dylib        	0x00000001b85bf1a4 __ulock_wait + 8
1   libdispatch.dylib             	0x00000001812ed064 _dlock_wait + 56 (lock.c:326)
2   libdispatch.dylib             	0x00000001812ed414 _dispatch_group_wait_slow + 60 (semaphore.c:214)
3   Logging                       	0x0000000106324dec -[DDLog lt_log:] + 101868 (DDLog.m:0)
4   Logging                       	0x00000001063235b4 __40-[DDLog queueLogMessage:asynchronously:]_block_invoke + 95668 (DDLog.m:325)
5   libdispatch.dylib             	0x00000001812ec670 _dispatch_client_callout + 20 (object.m:560)
6   libdispatch.dylib             	0x00000001812fb8c4 _dispatch_lane_barrier_sync_invoke_and_complete + 56 (queue.c:1016)
7   Logging                       	0x0000000106323560 -[DDLog queueLogMessage:asynchronously:] + 95584 (DDLog.m:336)
8   Logging                       	0x0000000106323b84 -[DDLog log:message:level:flag:context:file:function:line:tag:] + 97156 (DDLog.m:476)
9   Logging                       	0x0000000106323784 -[DDLog log:level:flag:context:file:function:line:tag:format:] + 96132 (DDLog.m:0)
```

**Comment by silverhammermba** [comment]
Both crashes happen early on in our app's lifecycle, in methods called from `application:didFinishLaunchingWithOptions:`

**Comment by ffried** [comment]
@silverhammermba can you trace the issue back to a certain `DDLogXYZ` statement? If so, is that a synchronous log call by any chance (level error or greater by default)?

**Comment by silverhammermba** [comment]
@ffried yes I can trace it back, and no both log statements are lower (former is info, latter is warn). We're on Cocoalumberjack 3.7.2

**Comment by ffried** [comment]
@silverhammermba Ok, thanks for the info! I'm still a bit confused why those calls would end up in a deadlock.
Maybe I can find some time to investigate this.

How reliably can you reproduce this? Would you be able to provide a sample project?

**Comment by randarnold** [comment]
I periodically see deadlocks as well. But only when stepping in the Xcode debugger. (I try to avoid stepping over any log statements.)

**Comment by silverhammermba** [comment]
@ffried I have not been able to reproduce this myself. I'm getting the stack traces from Apple's crash reporting. This is our  number 1 crash from customers installing our app in the app store :(

**Comment by silverhammermba** [comment]
I may have reproduced it. But I can only get it to happen when our app is built in release mode with no debugger attached. Tried reverting to Cocoalumberjack 3.1.0 (which we used to use) but it still crashed on app launch, only after locking up for 20 seconds or so.

**Comment by ffried** [comment]
@silverhammermba The 20 seconds make sense. It's the watchdog from iOS that terminates apps that hang.

I've checked your stack traces against our code and it definitively seems to log synchronously!

Is you application in Swift or Objective-C?

**Comment by silverhammermba** [comment]
@ffried our app is mixed swift and objective-c, this part it's crashing on is all objective-c.

I did find a workaround for the crash, still no root cause. I will try to put together a small app for reproducing it, though I'm a bit busy with rushing out a hotfix. Here is what our app is doing and what the workaround was:

In addition to Cocoalumberjack, we also use the `WS1IntelligenceSDK` pod for VMWare crash analytics (it sucks, don't give them money). The steps during launch in our app delegate are

1. Init WS1, give them the device UUID (this doesn't turn on crash reporting yet)
2. Init Cocoalumberjack, give it a log formatter which also passes every log message to our AppDelegate.
   * when the AppDelegate receives a log message, it forwards it to WS1 so that we see the latest logs in our crash analytics
3. Set up a notification observer on the main thread for WS1. If WS1 detects that the app crashed the last time it was used, WS1 sends us a notification. When we receive it, we log a warning and stop observing that notification.
4. Enable crash reporting in WS1

The workaround was to comment out the line in AppDelegate that forwards ours logs to WS1. In my mind the sequence of events is something like

1. App crashes for some reason
2. User restarts app
3. We get the WS1 notification that app crashed last time
4. Try to log a warning
5. Cocoalumberjack asks our log formatter to format it
6. Log formatter first tells AppDelegate about the log
7. AppDelegate passes the log to WS1
8. ????
9. deadlock

I can see how that is risky. Basically our log formatter blocks Cocoalumberjack, which may be in the middle of some mulithreading shenanigans, and starts doing stuff in a totally separate library. I bet I could fix this by making the log formatter use some async mechanism like a notification to pass the log to the AppDelegate instead. Still would be nice to know the real root cause.

**Comment by ffried** [comment]
> our app is mixed swift and objective-c, this part it's crashing on is all objective-c.

Do you somewhere change the define `LOG_ASYNC_ENABLED`? And are you using our default `DDLogXYZ` macros?


> Basically our log formatter blocks CocoaLumberjack, which may be in the middle of some multihreading shenanigans, and starts doing stuff in a totally separate library

This sound very much like a possibility for a deadlock. Especially since everything happens on the main thread. Also, CocoaLumberjack has a (serial) dispatch queue for each logger as well as a (serial) queue for all loggers, that is synced using a dispatch group.
The log formatter is called on the serial queue of the logger. Making a blocking call here might block the whole logging system. While loggers are generally given log messages in parallel, each log message is awaited to not overflow a single logger that is slower than others.

I'd recommend to implement your own logger for passing on log messages to another system. If passing them on is slow, maybe even maintain your own queue inside this logger for this particular step. Using a dedicated logger allows the message to be logged by other loggers in parallel instead of blocking them.

---

Do you have the contents of the other threads in your backtrace by any chance? This would maybe show, where the deadlock happens. Your backtrace above showing thread 0 doesn't seem to be the cause for the deadlock - just the one running into it.

**Comment by silverhammermba** [comment]
We don't mess with `LOG_ASYNC_ENABLED`. We don't directly call `DDLogXYZ`, we call our own macro which calls it, but I don't see any risky stuff there.

I checked the other stack traces we have and indeed all of them show some background thread in the middle of a `lt_log` call while thread 0 is stuck waiting on some sync mechanism. I also verified that our workaround was only a partial fix, we still have some users' apps crashing on log calls. I'm still a little confused how any of this can deadlock. The log calls that are blocking thread 0 are all async (DDLogInfo). Even if Cocoalumberjack is getting deadlocked internally, shouldn't those async logs immediately return?

We looked around for other mechanisms tied into logging and we similarly have another delegate call which is encrypting log messages. It sounds like that should also be a separate logger with its own internal queue. We'll try implementing that in the meantime.

**Comment by ffried** [comment]
@silverhammermba I'm probably as confused as you are 👀 
I've triple checked our code by now. If the log call was made using an async macro, the first interaction with a queue is `async` on the `_loggingQueue`. There is some syncing across queues internally later, but that first call is async and thus shouldn't block...

What I've noticed from your backtraces is that the line numbers don't match with version 3.7.2. Do you have backtraces that happened with version 3.7.2 by any chance? Just so that I can confirm it's really using the async path.

**Comment by silverhammermba** [comment]
🤦 if it's any consolation, you were completely right the whole time and we overlooked the obvious problem in our code.

Our code was only calling DDLogXYZ in Swift, in Objective-C we had a custom macro which was calling `-[log:level:flag:context:file:function:line:tag:format:,...]` and we had inverted the first argument. We thought the flag was for _synchronous_ logging, not asynchronous. Consequently almost all of our log messages were synchronous (and we have a _lot_ of log messages). Frankly I'm amazed our app wasn't crashing more.

My earlier workaround helped because it made each log slightly faster, which helped to mitigate all of the thread blocking.

So, at least for us, the RCA for this `lt_log` crash is way too much synchronous logging. Sorry for wasting your time, and thank you so much for your help.

Maybe in Cocoalumberjack 4.0 you can rename that method to `-[logAsync:level:flag:context:file:function:line:tag:format:,...]`, that would have made it much more obvious that we were doing a dumb thing.

**Comment by ffried** [comment]
@silverhammermba As awkward as this might be for you, I'm actually relieved. I was really starting to doubt everything here. 😅

I agree that the naming isn't great here. Maybe I'll just deprecated the old methods and add new versions of them. That way we don't have to wait until version 4.0 😉 

<hr/>

For anyone else with issue: Please first make sure you are **not** logging synchronously and then provide back traces for your crashes. They help a lot in diagnosing the underlying cause! Until then I'll close this issue.
